### PR TITLE
Fix fw boot retry - V3

### PR DESCRIPTION
**Pull request by ranj063**
This is a follow-up to https://github.com/thesofproject/linux/pull/507.

### CODE REVIEW

File: sound/soc/sof/intel/hda-loader.c [hunk 1]
```diff
@@ -306,24 +306,28 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+	if (ret < 0) {
 		dev_err(sdev->dev, "error: load fw failed ret: %d\n", ret);
-	else
-		dev_dbg(sdev->dev, "Firmware download successful, booting...\n");
+		goto err;
+	}
+
+	dev_dbg(sdev->dev, "Firmware download successful, booting...\n");
+	goto done;
+
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda-loader.c:343]: How about changing it like this:

```
	if (!ret) {
		dev_dbg(sdev->dev, "Firmware download successful, booting...\n");
		goto done;
	}
        dev_err(sdev->dev, "error: load fw failed ret: %d\n", ret);

err:
```

Comment (plbossart) [reply sound/soc/sof/intel/hda-loader.c:343]: +1

File: sound/soc/sof/intel/hda-loader.c [hunk 2]
```diff
@@ -306,24 +306,28 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+err:
+	hda_dsp_dump(sdev, SOF_DBG_REGS | SOF_DBG_PCI | SOF_DBG_MBOX);
+
+	/* disable DSP */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR,
+				SOF_HDA_REG_PP_PPCTL,
+				SOF_HDA_PPCTL_GPROCEN, 0);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c:351]: going from here to done: looks odd, why would you get the stream tag at this point?

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:351]: @plbossart @keyonjie I've fixed this to disable the DSP after stream cleanup.

File: sound/soc/sof/intel/hda-loader.c [hunk 3]
```diff
@@ -306,49 +321,38 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
 			i, ret);
-
-		hda_dsp_dump(sdev, SOF_DBG_REGS | SOF_DBG_PCI | SOF_DBG_MBOX);
-
-		/* disable DSP */
-		snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR,
-					SOF_HDA_REG_PP_PPCTL,
-					SOF_HDA_PPCTL_GPROCEN, 0);
-		goto done;
+		goto cleanup;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c:315]: you could also do the loop to find a stream before the error handling and jump to cleanup, that way you do it only once?

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:315]: @plbossart good idea. fixed now.

File: sound/soc/sof/intel/hda-loader.c [hunk 4]
```diff
@@ -306,49 +312,39 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
...
+	 */
 	ret1 = cl_cleanup(sdev, &sdev->dmab, stream);
 	if (ret1 < 0) {
 		dev_err(sdev->dev, "error: Code loader DSP cleanup failed\n");
-		return ret1;
+		ret = ret1;
 	}
 
+	if (!ret)
+		return ret;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader.c:342]: is there a reason why you want to return here without doing the DSP disable? Or is it a leftover from the previous code before you added the DSP disable?

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:342]: @plbossart we should disable the DSP only in case of errors. if ret is 0, it means fw boot was successful.

Comment (plbossart) [reply sound/soc/sof/intel/hda-loader.c:342]: @ranj063 the diff format is misleading, see the entire code

````c
ret1 = cl_cleanup(sdev, &sdev->dmab, stream);
if (ret1 < 0) {
	dev_err(sdev->dev, "error: Code loader DSP cleanup failed\n");			
        ret = ret1;
}


if (!ret)
	return ret;
````

Why would you return if ret1 was not zero? the cleanup failed at this point

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:342]: @plbossart if cleanup fails, ret1 will be negative and I set ret to ret1 and the condition (!ret) will not be true. So we will not return but proceed with disabling the DSP.

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:342]: @plbossart maybe I should add some comments to make it clearer

Comment (ranj063) [reply sound/soc/sof/intel/hda-loader.c:342]: @plbossart I've added some comments to make the error flow a bit clearer. Please let me know if thats helpful.

**Pull request merged**